src/eval.rs

```diff
@@ -215,27 +215,47 @@ impl Default for MiriConfig {
     }
 }
 
+struct MainThread<'tcx> {
```

Comment:
The name is too generic, given that this is specifically the main-thread-stack-empty-handler.

Also, all this data is only needed in `MainThreadState::GlobalCtors` right? They become dead weight afterwards. Please structure the enum to make it clear which data is actually needed when.